compiler/src/iree/compiler/DispatchCreation/FusionUtils.cpp

```diff
@@ -97,4 +101,35 @@ bool areFusableAsElementwiseOps(MLIRContext *context, OpOperand *fusedOperand,
   return true;
 }
 
+// Returns true when an operation `op` is horizontal to `currGroup` when
+// considering the program slice between `seedOp` and `op`.
+bool isHorizontalToGroup(Operation *op, ArrayRef<Operation *> currGroup,
+                         const DominanceInfo &dominanceInfo,
+                         Operation *seedOp) {
+  assert(dominanceInfo.properlyDominates(seedOp, op) &&
+         op->getParentRegion() == seedOp->getParentRegion());
+  BackwardSliceOptions options;
+  // Limit the slice to the seed to make sure the slice is small.
+  options.filter = [&](Operation *op) {
+    return !dominanceInfo.properlyDominates(op, seedOp);
+  };
+  llvm::SetVector<Operation *> slice;
+  getBackwardSlice(op, &slice, options);
+
+  // `getBackwardSlice` doesnt track uses from within an ops region, so make
```

Comment:
Not sure I follow this comment. Can you explain more?

Reply:
```mlir
%0 = linalg.generic
%1 = linalg.generic %0
%2 = linalg.generic %1 ins(%0 : tensor<...>) {
  ^bb0(%in: i64, %out: i64):
    %extracted = tensor.extract %1 [...]
    linalg.yield %extracted : f16
  } -> tensor<...>
```

In the above, the backward slice starting at `%2` would only include `%2` (but not `%1` because it isn't a direct operand). Which makes it seem like %2 can be moved before %1. Just looking at the slice doesn't tell you if there is a dependency between the ops.

This was causing issues with the open llama regression tests since (i think) there are multiple gathers

Reply:
Do we need to use a slice of the values that are captured from above to make sure that there is no dependence?

Reply:
Yeah I think so. Maybe this could be a option of `getBackwardsSlice`?

Reply:
Maybe... lets leave this this way for now.

Reply:
If I am seeing the test correctly there wasnt a test added for this? The failure in https://github.com/iree-org/iree/issues/18879 seems very related to this.

Reply:
@nirvedhmeshram There were several other tests that failed related to the backward slice problem (regression tests), but none were because the `consumer` was using values defined above which is why this slipped through. I'll make sure to add more testing when fixing this